azkaban-common/src/main/java/azkaban/executor/container/ContainerImplUtils.java

```diff
@@ -82,4 +92,65 @@ public static int getFlowNameHashValMapping(ExecutableFlow flow) {
     int flowNameHashVal = Math.abs(MurmurHash3.hash32x86(flowNameBytes));
     return flowNameHashVal % 100 + 1;
   }
-}
+
+  public static Set<String> getProxyUsersForFlow(final ProjectManager projectManager,
+      final ExecutableFlow flow) {
+    final Set<String> proxyUsers = new HashSet<>();
+    populateProxyUsersForFlow(flow, flow, projectManager, proxyUsers);
+    return proxyUsers;
+  }
+
+  public static void populateProxyUsersForFlow(final ExecutableFlow flow,
+      final ExecutableNode node, final ProjectManager projectManager, Set<String> proxyUsers) {
+    if (node instanceof ExecutableFlowBase) {
+      final ExecutableFlowBase base = (ExecutableFlowBase) node;
+      for (ExecutableNode subNode : base.getExecutableNodes()) {
+        populateProxyUsersForFlow(flow, subNode, projectManager, proxyUsers);
+      }
+    } else {
+      // If a node is disabled, we don't need to initialize its jobType container image when
+      // creating a pod.
+      if (node.getStatus() != Status.DISABLED) {
+        Project project = projectManager.getProject(flow.getProjectId());
+        Flow flowObj = project.getFlow(flow.getFlowId());
+        Props currentNodeProps = projectManager.getProperties(project, flowObj,
+            node.getId(), node.getJobSource());
+        // Get the node level property for proxy user.
+        String userToProxyFromNode = currentNodeProps.getString(USER_TO_PROXY, null);
+        // Get the node level override by user from the UI for proxy user.
+        Props currentNodeJobProps = projectManager.getJobOverrideProperty(project, flowObj,
+            node.getId(), node.getJobSource());
+        String userToProxyFromJobNode = currentNodeJobProps.getString(USER_TO_PROXY, null);
+        if (userToProxyFromJobNode != null) {
+          proxyUsers.add(userToProxyFromJobNode);
+        } else if (userToProxyFromNode != null) {
+          proxyUsers.add(userToProxyFromNode);
+        }
+      }
+    }
+  }
+
+  // Extract the proxy users needed from  PREFETCH_JOBTYPE_PROXY_USER_MAP
+  // This method is being introduced to be able to assign a specific proxy user that will require
+  // custom credentials for a given job type. This allows the jobtype to perform specific checks
+  // without requiring azkaban executor's credentials to do this, once we enforce POLP defined in
+  // https://github.com/azkaban/azkaban/pull/3216
+  // This will parse the jobTypePrefetchUserMap of the format : "jobtype1,jobtype1_proxyuser;
+  // jobtype2,jobtype2_proxyuser" and add the proxy user for a given job if that jobtype is
+  // present in the flow.
```

Comment:
Thanks for adding the comment!
nit: Could you make it a method doc comment by replacing `//` with `/**...*/`

Also, I'm wondering, will the jobs be able to impersonate the jobtypes proxy users? How is that prevented?

Reply:
Will change this thanks !
potentially yes. if they break out of the job process. However, the impact will be limited to access to the other users prefetched, which is an improvement over the current situation of the azkaban executor user.